module_utils/network_lsr/argument_validator.py

```diff
@@ -413,6 +414,73 @@ def _validate_post(self, value, name, result):
 
         return result
 
+class ArgValidatorIPRule(ArgValidatorDict):
+    def __init__(self, name, family=None, required=False, default_value=None):
+        ArgValidatorDict.__init__(
+            self,
+            name,
+            required,
+            nested=[
+                ArgValidatorIP(
+                    "from", family=family, required=False, plain_address=False
+                ),
+                ArgValidatorNum("from_prefix", default_value=32, val_min=0, val_max=32),
+                ArgValidatorIP(
+                    "to", family=family, required=False, plain_address=False
+                ),
+                ArgValidatorNum("to_prefix", default_value=32, val_min=0, val_max=32),
+                ArgValidatorStr(
+                    "iif", default_value=None, enum_values=listdir('/sys/class/net/')
+                ),
+                ArgValidatorStr(
+                    "oif", default_value=None, enum_values=listdir('/sys/class/net/')
+                ),
+                ArgValidatorNum(
+                    "tos", default_value=None, val_min=0, val_max=255
+                ),
+                ArgValidatorNum(
+                    "dsfield", default_value=None, val_min=0, val_max=255
+                ),
+                ArgValidatorNum(
+                    "fwmark", default_value=None, val_min=0, val_max=65535
+                ),
+                # Not implemented
+                # ArgValidatorRange(
+                #     "uidrange", default_value=None, val_min=0, val_max=65535
+                # ),
+                ArgValidatorNum(
+                    "ipproto", default_value=None, val_min=0, val_max=255
+                ),
+                # ArgValidatorRange(
+                #     "sport", default_value=None, val_min=0, val_max=65535
+                # ),
+                # ArgValidatorDport(
+                #     "dport", default_value=None, val_min=0, val_max=65535
+                # ),
+                ArgValidatorNum(
+                    "priority", default_value=None, val_min=0, val_max=32767
+                ),
+                ArgValidatorStr(
+                    "table", default_value="main"
+                ),
+                ArgValidatorNum(
+                    "protocol", default_value=None, val_min=1, val_max=255
```

Comment:
NetworkManager supports 20 rule attributes, "protocol" is not one of them.

That makes me wonder whether this is actually useful. Do you have an actual use-case for this? I think we should add features carefully as they are used. In particular, because the patch does not add other attributes which I think might be more useful. Let's start small, and add features as they are used.

Reply:
No, I don't.

In ip rules, I implemented everything what was in man ip-rule. The other approach I could have had is to implement only attributes I need, which would not be totally fair to possible other users of the role.

But I agree with removing not really used attributes and add just some basic.

Reply:
Btw, in our use case we need to use `uidrange` and `sport` parameters, which are not supported below iproute2 4.15 and I don't exactly now which kernel version, but definitely not in 7.x's 3.10, and present in 8.x's 4.18. How should I treat this?

Reply:
They can still be added.

Possibly `run_prepare()` could detect such missing features and either fail (fatal) or warn about it. In practice it's however not easy to understand whether the feature is available or not...